File: 06.zigzag-conversion/zigzag.go
package main

import (
	"bytes"
	"fmt"
)

func zigzagConversion(text string, nRows int) string {
	if nRows <= 1 || len(text) <= nRows {
		return text
	}

	res := bytes.Buffer{}
	pace := 2*nRows - 2

	// process first row
	for i := 0; i < len(text); i += pace {
		res.WriteByte(text[i])
	}

	// process 2nd ... last-1 row
	for row := 1; row <= nRows-2; row++ {
		res.WriteByte(text[row])
		for k := pace; k-row < len(text); k += pace {
			res.WriteByte(text[k-row])
			if k+row < len(text) {
				res.WriteByte(text[k+row])
			}
		}
	}

	// process last row
	for i := nRows - 1; i < len(text); i += pace {
		res.WriteByte(text[i])
	}
	return res.String()
}

func solveProblem(text string, nRows int) {
	res := zigzagConversion("PAYPALISHIRING", 3)
	fmt.Printf("string %s, row: %d, result: %s\n", text, nRows, res)
	if res == "PAHNAPLSIIGYIR" {
		fmt.Println("Conversion is correct")
	} else {
		fmt.Println("Failed  to zigzag convert")
	}
}

func main() {
	solveProblem("PAYPALISHIRING", 3)

}

File: 06.zigzag-conversion/zigzag.py

from py_compile import main
from socket import if_nameindex


def zigzag_conversion(text: str, row_number: int) -> str:
    if len(text) <= 1 or len(text) <= row_number:
        return text

    res = []
    pace = 2 * row_number - 2

    # first row
    for i in range(0, len(text), pace):
        res.append(text[i])

    # 2nd ... last - 1 row
    for row in range(1, row_number - 1):
        res.append(text[row])
        for k in range(pace, len(text), pace):        
            res.append(text[k - row])
            if k + row < len(text):
                res.append(text[k + row])

    # last row
    for i in range(row_number - 1, len(text), pace):
        res.append(text[i])

    return "".join(res)


def solve_problem(text: str, row_number: int) -> None:
    res = zigzag_conversion(text, row_number)
    print(f"string: {text}, row: {row_number}, result: {res}")
    if res == "PAHNAPLSIIGYIR":
        print("zigzag conversion was correct")
    else:
        print("Failed to convert")

if __name__ == '__main__':
    solve_problem("PAYPALISHIRING", 3)
